ledger/simulation/tracer.go

```diff
@@ -367,31 +378,58 @@ func (tracer *evalTracer) AfterOpcode(cx *logic.EvalContext, evalError error) {
 		}
 	}
 
-	if cx.RunMode() != logic.ModeApp {
-		// do nothing for LogicSig ops
-		return
+	if cx.RunMode() == logic.ModeApp {
+		tracer.handleError(evalError)
+		if evalError == nil && tracer.unnamedResourcePolicy != nil {
+			if err := tracer.unnamedResourcePolicy.assignment.reconcileBoxWriteBudget(cx.BoxDirtyBytes(), cx.Proto.BytesPerBoxReference); err != nil {
+				// This should never happen, since we limit the IO budget to tracer.unnamedResourcePolicy.assignment.maxPossibleBoxIOBudget
+				// (as shown below), so we should never have to reconcile an unachievable budget.
+				panic(err.Error())
+			}
+
+			// Update box budget. It will decrease if an additional non-box resource has been accessed.
+			cx.SetIOBudget(tracer.unnamedResourcePolicy.assignment.maxPossibleBoxIOBudget(cx.Proto.BytesPerBoxReference))
+		}
 	}
-	tracer.handleError(evalError)
 }
 
 func (tracer *evalTracer) BeforeProgram(cx *logic.EvalContext) {
 	groupIndex := cx.GroupIndex()
 
 	// Before Program, activated for logic sig, happens before txn group execution
 	// we should create trace object for this txn result
-	if cx.RunMode() != logic.ModeApp {
+	if cx.RunMode() == logic.ModeSig {
 		if tracer.result.ReturnTrace() {
 			tracer.result.TxnGroups[0].Txns[groupIndex].Trace = &TransactionTrace{}
 			traceRef := tracer.result.TxnGroups[0].Txns[groupIndex].Trace
 			traceRef.programTraceRef = &traceRef.LogicSigTrace
 		}
 	}
+
+	if cx.RunMode() == logic.ModeApp && tracer.unnamedResourcePolicy != nil {
+		globalSharing := false
+		for iter := cx; iter != nil; iter = iter.GetCaller() {
+			if iter.ProgramVersion() >= 9 {
+				// If some caller in the app callstack allows global sharing, global resources can
+				// be accessed here. Otherwise the top-level txn must declare all resources locally.
+				globalSharing = true
+				break
+			}
+		}
```

Comment:
We discussed this live once - I don't know that we came to a good conclusion.  There's a sense in which this true - if there is a v9 or higher caller, then the resource could be passed down, in principle.

But the reality is that, if it wasn't passed down, there's no reason to expect that the main strategy underlying the whole "simulate, then pack, then execute" is going to work to get it passed down.  How could the packer possibly know how to convince the calling code to do it?

I'm perfectly willing to believe that a v9 caller might be smart enough to know how to call a v8 program by packing the necessary resources, but I think that would happen if, for example, it obtained the resource from some argument (or state), and then knew that since it was calling a v8 program, it should pack it. So in a simulate call, we would observe the v9 caller accessing the resource, and then we would see the v8 program using properly "named" access because the caller _will_ pack the resource, even in simulation. Or else, I think, nothing will ever convince it to do so.

So I think either:
1. a v8 *inner* program should _fail_ even in simulation and with unnamed resource access turned on if it accesses unnamed resources
2. or, we at least need to note such a "deep" unnamed access by pre v9 in the response and let the "calling" wallet conclude the transaction won't work

I suppose you are indeed doing #2 already - your assignment structure comes back for the inner with a needed local access?

Reply:
Let's consider the 3 main uses cases this PR will enable:
1. Smart contract developers who are debugging their code
2. Read-only off-chain callers
3. Wallets/users who want to prep transactions with foreign references before sending them

The 3rd use case is the only one concerned with producing a valid transaction that can run under normal circumstances. The other two can still operate and be useful even if a v8 inner program accesses an unnamed resource, so I consider this an argument against implementing your first suggestion.

Your second suggestion seems reasonable though, as it would give callers an indication that even if they pack every resource appropriately, their txn will still fail. However, I'll argue that this is more of a nice-to-have, since without it they'll still see a failure, just at submission time. Furthermore, as you mentioned during standup, this technique would produce a false positive if there are any programs that "forward" their foreign ref arrays to inner calls.

So for the time being I'm arguing against any changes here. I don't think the added benefit to failing faster is worth the risk of alienating programs that can work properly. I think this is definitely a shortcoming we should document when we work on #5616 though.

Reply:
I buy your argument against the first idea.

As for the second, perhaps you're already doing what I said?  You said you push the unnamed access objects up to the top-level transaction, so that a top-level transaction carries info about all of the things its callees touched, right?  But do you _also_ return the unnamed accesses of the inners along with those inners. If so, I think that's all we can do.  You'd be returning the fact that this program needs something it doesn't have.

I supposed maybe you don't do that. Since you consider globalSharing to be true in that case, it only appears at the top-level, as something that can be addressed by group sharing?

Can I argue for both, in a sense?  The inner v8 program should still return what it touched, because it _can't_ be fixed (directly) by group sharing. But the top-level transactions should also indicate that it needs the resource and _can_ be fixed by group sharing.

Reply:
> I supposed maybe you don't do that. Since you consider globalSharing to be true in that case, it only appears at the top-level, as something that can be addressed by group sharing?

Correct, there is no individual reporting of inner txn resources. Everything is either in the global scope, or (when resource sharing is not possible) in the top-level txn scope.

Imagining if we also reported inner txn resources separately, can you share how you see this being used? If your answer is that a wallet will raise an error if it sees any v8 inner resources accessed, I'm not convinced by this.

Reply:
Ok, I don't have anything constructive to add, and I don't think this is so bad. If dealing with v8 programs called by v9 programs is a little non-obvious, so be it.